--- hypre-2.31.0/src/utilities/memory.c	2024-02-14 12:17:26.000000000 -0700
+++ hypre-2.31.0.patched/src/utilities/memory.c	2024-09-27 11:00:26.205878311 -0600
@@ -1136,7 +1136,7 @@
 #if defined(HYPRE_USING_HIP)
 
    struct hipPointerAttribute_t attr;
-   *memory_location = hypre_MEMORY_UNDEFINED;
+   *memory_location = hypre_MEMORY_HOST;
 
    hipError_t err = hipPointerGetAttributes(&attr, ptr);
    if (err != hipSuccess)
