diff --git a/algorithms/src/Kokkos_Random.hpp b/algorithms/src/Kokkos_Random.hpp
index c6fe04e9c..5963c15f1 100644
--- a/algorithms/src/Kokkos_Random.hpp
+++ b/algorithms/src/Kokkos_Random.hpp
@@ -895,13 +895,14 @@ class Random_XorShift64 {
   // number
   KOKKOS_INLINE_FUNCTION
   double normal() {
-    constexpr auto two_pi = 2 * Kokkos::numbers::pi_v<double>;
-
-    const double u     = drand();
-    const double v     = drand();
-    const double r     = Kokkos::sqrt(-2.0 * Kokkos::log(u));
-    const double theta = v * two_pi;
-    return r * Kokkos::cos(theta);
+    double S = 2.0;
+    double U;
+    while (S >= 1.0) {
+      U              = 2.0 * drand() - 1.0;
+      const double V = 2.0 * drand() - 1.0;
+      S              = U * U + V * V;
+    }
+    return U * std::sqrt(-2.0 * std::log(S) / S);
   }
 
   KOKKOS_INLINE_FUNCTION
@@ -1168,13 +1169,14 @@ class Random_XorShift1024 {
   // number
   KOKKOS_INLINE_FUNCTION
   double normal() {
-    constexpr auto two_pi = 2 * Kokkos::numbers::pi_v<double>;
-
-    const double u     = drand();
-    const double v     = drand();
-    const double r     = Kokkos::sqrt(-2.0 * Kokkos::log(u));
-    const double theta = v * two_pi;
-    return r * Kokkos::cos(theta);
+    double S = 2.0;
+    double U;
+    while (S >= 1.0) {
+      U              = 2.0 * drand() - 1.0;
+      const double V = 2.0 * drand() - 1.0;
+      S              = U * U + V * V;
+    }
+    return U * std::sqrt(-2.0 * std::log(S) / S);
   }
 
   KOKKOS_INLINE_FUNCTION