--- a/include/systems/parameter_accessor.h	2022-08-08 08:34:15.000000000 -0700
+++ b/include/systems/parameter_accessor.h	2025-11-04 13:18:49.468632000 -0800
@@ -112,7 +112,7 @@
   /**
    * Setter: change the value of the parameter we access.
    */
-  ParameterProxy & operator = (const ParameterProxy<T> & new_value) { _accessor.set(new_value.get()); }
+  ParameterProxy & operator = (const ParameterProxy<T> & new_value) { _accessor.set(new_value._accessor.get()); return *this; }
 
   /**
    * Setter: change the value of the parameter we access.
@@ -144,6 +144,11 @@
    */
   operator T () const { return _accessor.get(); }
 
+  /**
+   * Getter: get the value of the parameter we access.
+   */
+  T get() const { return _accessor.get(); }
+
 #ifdef LIBMESH_DEFAULT_QUADRUPLE_PRECISION
   operator boost::multiprecision::backends::float128_backend () const { return _accessor.get().backend(); }
 #endif